### PR TITLE
if there's an exception in the Client async context manager body then close fast

### PR DESCRIPTION
**Pull request by graingert**
Closes #xxxx

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/client.py [hunk 1]
```diff
@@ -1383,7 +1383,7 @@ async def __aenter__(self):
         return self
 
     async def __aexit__(self, exc_type, exc_value, traceback):
-        await self._close()
+        await self._close(fast=exc_type is not None)
```
Comment (graingert) [inline distributed/client.py:1386]: this needs a test

**Comment by graingert** [comment]
if anyone needs the old behaviour you can use a `shield_from_exception` context manager to transport the exception around the context manager:

```python
import sys
import contextlib
from distributed import Client
import asyncio


@contextlib.asynccontextmanager
async def shield_from_exception(cmgr_factory, /, *args, **kwargs):
    err = None
    async with cmgr_factory(*args, **kwargs) as v:
        try:
            yield v
        except BaseException as e:
            err = e

    if err is not None:
        raise err


async def amain():
    async with shield_from_exception(Client, asynchronous=True) as client:
        raise Exception

def main():
    return asyncio.run(amain())

if __name__ == "__main__":
    sys.exit(main())
```

File: distributed/tests/test_client.py [hunk 2]
```diff
@@ -7571,3 +7572,21 @@ def __init__(self, *args, **kwargs):
...
     ]
+
+
+@gen_cluster(client=False, nthreads=[])
+async def test_fast_close_on_aexit_failure(s):
+    class MyException(Exception):
+        pass
+
+    c = Client(s.address, asynchronous=True)
+    with mock.patch.object(c, "_close", wraps=c._close) as _close_proxy:
```
Comment (graingert) [inline distributed/tests/test_client.py:7583]: I couldn't find any tests that tested what `_close(fast=True)` does.

File: distributed/tests/test_client.py [hunk 3]
```diff
@@ -7571,3 +7572,21 @@ def __init__(self, *args, **kwargs):
...
+        with pytest.raises(MyException):
+            async with c:
+                start = time()
+                raise MyException
+        stop = time()
+
+    assert _close_proxy.mock_calls == [mock.call(fast=True)]
+    assert c.status == "closed"
+    assert (stop - start) < 2
```
Comment (graingert) [inline distributed/tests/test_client.py:7592]: I couldn't find any tests that tested what `_close(fast=True)` does:

it just reduces this 2 second timeout to 0
https://github.com/dask/distributed/blob/c373746a1fae0803be484474042bfdc18ab01699/distributed/client.py#L1584

**Review by hendrikmakait — COMMENTED** [review]
Code looks good to me, could you elaborate on the motivation for this change?

File: distributed/tests/test_client.py (hunk 3 above)
Comment (graingert) [reply distributed/tests/test_client.py:7559]: introduced here https://github.com/dask/distributed/pull/27/files#diff-ce261e4d4b0185c8defbfcd46ebe8311a3ecb7ab81d653a2c6664c840340ab2cR250-R251

Comment (graingert) [reply distributed/tests/test_client.py:7559]: which is this: https://trio.readthedocs.io/en/stable/reference-io.html#trio.aclose_forcefully

Comment (graingert) [reply distributed/tests/test_client.py:7559]: @hendrikmakait thanks for the review, it helped clarify my thinking on this!

Comment (hendrikmakait) [reply distributed/tests/test_client.py:7559]: nit: Would it make sense to rename `fast` to `force`?

**Review by hendrikmakait — APPROVED** [review]
LGTM, thanks for adding the explanation. Feel free to ignore my nit, this just crossed my mind after reading your clarification.

File: distributed/tests/test_client.py (hunk 3 above)
Comment (graingert) [reply distributed/tests/test_client.py:7559]: Yeah I think "forcefully=True" is probably the best here

**Review by fjetter — APPROVED** [review]

**Pull request merged**